### PR TITLE
Runtime Support for new parallel type

### PR DESCRIPTION
**Pull request by samnordmann**
# What's in this patch:
- support Fusion with tensor which are sharded across devices (with the new parallel type DiDx), with the following restrictions:
  -  restrictions described in #918, basically, tensors can only be shared on their first dimension, on 1D device meshes. This implies that tensors are either fully sharded or fully replicated
  - For now, the Pipeline has to be set manually, in a way that there is no communication besides the inter-stages communications.
  - The only allowed Expr at inter-stage is `Set`
- We add testing utilities to compare the results of a Fusion executed on a single device with the multi-device version. The tests are organized to be modular, minimize code duplication, and improve performance

# Implementation details and motivations:
- **Runtime:**
  -  **Use `lowerCommunication` in the executor**. See the function `void PipelineExecutor::handle(PipelineCommunication* c)` in [multidevice/executor.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-d783a79fdc189059065a81ba9acb6d6c3d3ecc5c70188706febc97d7fea5aa36). We remove the old implementation, and instead, we call the function `lowerCommunication` (introduced recently in #918) that transforms the `PipelineCommunication` into a series of `Communication`. Then we `post` and `wait` these `Communications`
  - **Add Option to disable auto-scheduling**.
     - We add in `PipelineStage` an attribute `bool auto_schedule` (see [multidevice/pipeline.h](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-e66fbfbcfc303fec7ffb2c82e1f130dd7b72039a5601f8a146e2a0681824c725)). Before, we always used auto-scheduling for each different stage -- but now we want to manually set the parallel type and so we want to be able to turn off the auto-scheduling.
     - We thus handle this bool in the function `void PipelineExecutor::handle(PipelineStage* c)` in [multidevice/executor.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-d783a79fdc189059065a81ba9acb6d6c3d3ecc5c70188706febc97d7fea5aa36), where we either use `FusionExecutor` or `FusionExecutorCache` depending on `bool auto_schedule`.
     - We need to implement `FusionExecutor::allocOutputSpace`, which is done in [csrc/executor.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-e72d395ea8a3ab3d2a909dfb3677beac1bbd89336b54ee858867f652b48be4f0)
   - **error handling in MultiDeviceRuntime:** small change in `MultiDeviceRuntime::validate()`, inside [multidevice/runtime.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-476e4d39a0bfe025d0030a81ee4a99d7d39c950fa096d109e2b4e2ee89d861fb). Before, this function was asserting, now it returns a string with an error msg. Empty string means no error. This is useful so that in the test we can handle the error as we want, notably we may want to simply skip the test if there are not enough ranks, instead of throwing an error.
- **Test:**
  - **Share the `Communicator` across all tests:** this is needed for two reasons: 1) creating the `Communicator` is a very costly operation, and 2) we cannot create multiple `Communicators` on the same `address:port` (this restriction comes from pytorch's ProcessGroups). We point out that defining a global static Communicator is not valid and often leads toa segfault during cleanup. Instead, To achieve this, we define the class `class MultiDeviceEnvironment : public testing::Environment` and add an instance to the GTEST environment with the command `testing::AddGlobalTestEnvironment(new MultiDeviceEnvironment))`. See [test/multidevice.h](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-0d4c54d65e1463bf971ea74222f9653194a9fff61015202ed105b01883d76b2b) and [test/multidevice.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-00a0e9da3365a4a923933dd9cf701175f28c5c8f157741884d5242ad0a02a49d)
  - **Tests utilities and structure:** we define different classes and methods in [test/multidevice.h](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-0d4c54d65e1463bf971ea74222f9653194a9fff61015202ed105b01883d76b2b). The structure we provide allow to avoid a lot of code duplication in the tests, and to make them more adaptable with future developments. Among the utilities we provide, `PipelineTest::validate()` allows to run a pipeline on a multidevice setting and to compare the results with the output of the same fusion executed on a single device.
  - **New tests**: We provide new tests using pipelines and the new parallel types in [test/test_multidevice_pipeline.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-ccc7195e288586558c5f7aff6761ba0f5825de3ce39e469b0858e86b3fa737b1) (this file has been renamed and was previously named [test/test_multidevice.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-9c5f8ee781e2c6ee3eb63e8f628becfab0c0514d5e0a21266811a76eabfda254))
  - **Clean the implementations of the old tests** in [test/test_multidevice_communications.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-b8f1ce82763ea4f5fb61b4773a1f4232e1325fa63b63616c7ffee7c0a48b2296)` by leveraging the new classes from [test/multidevice.h](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-0d4c54d65e1463bf971ea74222f9653194a9fff61015202ed105b01883d76b2b)
  - Remove a now obsolete check in [test/main.cpp](https://github.com/NVIDIA/Fuser/pull/1022/files#diff-7afe085f03fa2399a9d1d962f6aaa46c84e72300d791c4aa657f56cc70f638e0).

### CODE REVIEW

File: test/test_multidevice_pipeline.cpp
Comment (naoyam) [inline test/test_multidevice_pipeline.cpp:1]: I assume this file is copied from test_multidevice.cpp. Are there any real changes?

**Comment by naoyam** [comment]
Could you give us some hints to review the PR? Looks like the changes are all over, but most of them seem relatively small. I got an impression from the PR title that there would be some major changes in `runtime.h/cpp`, but that doesn't seem to be the case. We could talk more details tomorrow.

**Comment by samnordmann** [comment]
> Could you give us some hints to review the PR? Looks like the changes are all over, but most of them seem relatively small. I got an impression from the PR title that there would be some major changes in `runtime.h/cpp`, but that doesn't seem to be the case. We could talk more details tomorrow.

Sure! First, to help review, you can review each commit separately. You can also read the PR description, in which I added a lot more details. Let's also talk about it the next time we meet

**Comment by samnordmann** [comment]
!build

File: test/test_multidevice_pipeline.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        all_meshes,
+        all_meshes,
+        ::testing::Values(false),
+        ::testing::Values(false)));
+
+INSTANTIATE_TEST_SUITE_P(
+    Bcast_sharded,
+    PipelineTestTwoStages,
+    ::testing::Combine(
+        ::testing::Values(mesh2),
```
Comment (cowanmeg) [inline test/test_multidevice_pipeline.cpp:178]: What is this testing? It looks like no communication is necessary since tv1 and tv2 have the same mesh and sharding.

Comment (samnordmann) [reply test/test_multidevice_pipeline.cpp:178]: Well spotted thank you. It's a mistake, it needs to be two different meshes of same size. Let me fix it

File: csrc/multidevice/executor.cpp [hunk 2]
```diff
@@ -31,89 +28,71 @@ void PipelineExecutor::handle(PipelineStage* stage) {
...
-      for (size_t counter = 0; counter < nbr_dests_per_comm + (i < remainder);
-           counter++, j++) {
-        auto dst = receivers.at(j);
-        communication.team.push_back(dst);
-      }
-      communications.push_back(communication);
-    }
+  at::Tensor input_tensor =
+      val_to_IValue_.at(c->in())
+          .toTensor(); // when stage has same input and ouput
```
Comment (samnordmann) [inline csrc/multidevice/executor.cpp:73]: remove comment

File: csrc/multidevice/pipeline.cpp [hunk 3]
```diff
@@ -152,9 +152,6 @@ class PipelineBuilder final {
         // then add the newly created PipelineVal as an input of the pipeline
         if (isGlobalInput(val)) {
           pipeline_->addInput(p_val);
-          NVF_ERROR(
```
Comment (naoyam) [inline csrc/multidevice/pipeline.cpp:155]: Why is this removed?

File: csrc/multidevice/pipeline.cpp [hunk 4]
```diff
@@ -264,6 +261,16 @@ std::unique_ptr<Fusion> Pipeline::stageToFusion(PipelineStage*& stage) const {
   return fusion_copy;
 }
 
+uint64_t Pipeline::requestedNumberOfDevices() const {
```
Comment (naoyam) [inline csrc/multidevice/pipeline.cpp:264]: Maybe `requestedNumberOfUniqueDevices`?

File: csrc/multidevice/runtime.h [hunk 5]
```diff
@@ -25,9 +25,7 @@ namespace nvfuser {
 class MultiDeviceRuntime {
  public:
   explicit MultiDeviceRuntime(Pipeline* pipeline, Communicator& comm)
-      : pipeline_(pipeline), comm_(comm) {
-    validate();
```
Comment (naoyam) [inline csrc/multidevice/runtime.h:29]: Why is validation moved?

**Review by naoyam — COMMENTED** [review]
Just some minor comments for now

File: csrc/multidevice/runtime.h (hunk 5 above)
Comment (samnordmann) [reply csrc/multidevice/runtime.h:29]: Indeed, I moved the call to `validate` from the constructor to the method `runWithInput`. Why is it a problem? Why not do so? We can consider that It is always valid to virtually define the runtime, but possibly invalid to run it with concrete input.
This way, the user can instantiate the `MultiDeviceRuntime`, and then call `validate` to query whether the runtime is valid. This is done in the Gtests, where we want to skip the test rather than throwing an error if the runtime setup is not compatible with the Pipeline.

File: csrc/multidevice/pipeline.cpp (hunk 4 above)
Comment (samnordmann) [reply csrc/multidevice/pipeline.cpp:264]: ok

File: csrc/multidevice/pipeline.cpp (hunk 3 above)
Comment (samnordmann) [reply csrc/multidevice/pipeline.cpp:155]: we don't need this restriction. We want the inputs to be possibly sharded

File: csrc/multidevice/pipeline.h [hunk 6]
```diff
@@ -63,8 +63,11 @@ class PipelineStageDescriptor final {
   using ValSet = VectorOfUniqueEntries<Val*>;
 
  public:
-  PipelineStageDescriptor() : unique_id(running_unique_id_++) {}
+  PipelineStageDescriptor()
+      : auto_schedule(true), unique_id(running_unique_id_++) {}
 
+  // whether the stage should be auto-scheduled
+  bool auto_schedule;
```
[the remark this answers is not in the document]
Comment (samnordmann) [reply csrc/multidevice/pipeline.h:70]: We could, but what is the reason to do so? `PipelineStageDescriptor` is supposed to be fully configurable and be passed as argument for creating PipelineStage. We could still set it in the constructor but it will just make the syntax of writing tests a bit less flexible.
I think in the future this class will be renamed `PipelineStageParams` and be made a `struct` rather than a `class` (as you once suggested). I think it is more convenient to let the members public and non-const so they can be manually set in the tests

File: csrc/multidevice/communicator.h [hunk 7]
```diff
@@ -102,9 +102,9 @@ class Communicator {
   bool is_available_;
   CommunicatorBackend backend_type_;
   RankType rank_;
-  int64_t size_;
+  uint64_t size_;
```
[the remark this answers is not in the document]
Comment (samnordmann) [reply csrc/multidevice/communicator.h:105]: ok

File: test/multidevice.cpp [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+#include <test/validator.h>
+#include <torch/cuda.h>
+
+namespace nvfuser {
+
+auto multidevice_env = static_cast<MultiDeviceEnvironment*>(
+    testing::AddGlobalTestEnvironment(new MultiDeviceEnvironment));
+
+void MultiDeviceEnvironment::SetUp() {
+  communicator_ = new Communicator();
```
Comment (zasdfgbnm) [inline test/multidevice.cpp:23]: Does it make sense to use `std::unique_ptr`?

Comment (samnordmann) [reply test/multidevice.cpp:23]: In my opinion, this is not suitable since the pointer is used by all the instances of `MultiDeviceTest`, so we would have to pass the raw pointer each time. A shared pointer would work though.
Lmk whether you want me to make the change or leave like that.

Comment (zasdfgbnm) [reply test/multidevice.cpp:23]: Could you just use `unique_ptr` for managing ownership, and pass raw pointers for non-owning functions and classes? `shared_ptr` is for shared ownership, which I don't think is related here.

File: test/multidevice.cpp [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+      obtained.equal(expected),
+      "Device ",
+      communicator->deviceId(),
+      " expected tensor:\n",
+      expected,
+      "\nbut obtained tensor:\n",
+      obtained);
+}
+
+void CommunicationTest::resetDstBuffers() {
```
Comment (zasdfgbnm) [inline test/multidevice.cpp:59]: Why do we need to reset buffers? Does some code rely on the buffer being zero-inited? If not, could we fill `nan` instead of zeros, as it is more likely to uncover bugs?

File: csrc/multidevice/pipeline.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/multidevice/pipeline.h:70]: Ignoring the convenience of writing tests, can this be `const`? Making a const variable `const` is also helpful to understand the intention of the API design, which is why I asked this question.

File: csrc/multidevice/runtime.h (hunk 5 above)
Comment (naoyam) [reply csrc/multidevice/runtime.h:29]: OK, makes sense.

File: test/multidevice.cpp (hunk 9 above)
Comment (samnordmann) [reply test/multidevice.cpp:58]: Makes sense. I changed to it fill the tensors with nan

**Comment by samnordmann** [comment]
!build

**Comment by samnordmann** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM
